### DISCUSSION TITLE
Dynamically Add Menu to File Menu

### DISCUSSION THREAD

**Discussion by maheshrajrp** [eclipse-theia/theia#10157] [General] [answered]
Hi,
I'm trying to implement Recent Files Functionality by storing in preference. However, I'm unable to update the submenu present in Files dynamically by using the onPreferenceChanged event listener. Can we dynamically change menus in the menubar using similar event listeners ? If yes can provide some examples.

Thanks

**Comment by vince-fugnitto** [comment, marked as answer]
@maheshrajrp given that you are not updating the top-level menu item itself, it should be pretty straightforward. What have you tried so far?
You can also take a look at this past discussion which looks similar to your use-case: https://github.com/eclipse-theia/theia/discussions/8411.

**Reply by maheshrajrp** [reply]
Hi @vince-fugnitto ,
According to that discussion, browser isn't supported and have to reload for it to reflect but I'm hoping for a solution on browser. I'm currently registering both the commands and  MenuAction on the fly by using onPreferenceChanged event listener. But, ui doesn't reflect the changes.

**Reply by vince-fugnitto** [reply]
@maheshrajrp if you're referring to this comment (https://github.com/eclipse-theia/theia/discussions/8411#discussioncomment-54202) it is only for the top-level menu items and not submenus like you've presented in your use-case.

**Reply by maheshrajrp** [reply]
@vince-fugnitto Sorry, my bad. In that case, I'll try to replicate the same and get back if I encounter any issues. 

Thanks @vince-fugnitto .

**Reply by maheshrajrp** [reply]
Hi @vince-fugnitto ,
It works. However wont it take a hit on performance ? The menu is being reconstructed every time  I open a new project. Can we have a feature similar to this #9886 where we add/remove commands on the go and the UI can reflect it ?

**Comment by maheshrajrp** [comment]
Hi @vince-fugnitto ,
Also if possible do you have any other suggestion that might be more performance focused like instead of disposing the entire menu and recreating a new one, can't we make few tweaks to the existing ones ? Similar to the ones provided in "Open Recent" in VSCode.

**Reply by vince-fugnitto** [reply]
@maheshrajrp `open recent...` in vscode is not a top-level menu item. As far as I know, the only time we need to dispose the menu would be when we update the top-level menu. I'm not sure what your use-case is but from initial discussions it looks like you only wanted to update the submenu dynamically.

**Reply by maheshrajrp** [reply]
@vince-fugnitto My bad sorry. I wanted to replicate similar to that of VSCode's Open Recent... My usecase is that only.

**Reply by vince-fugnitto** [reply]
@maheshrajrp then there should not be a need to dispose and re-create the whole menu.

**Reply by maheshrajrp** [reply]
@vince-fugnitto 
But still using onPreferenceChange Emitter doesn't add to the existing Menu. I had to refresh every time to get the change.  Can you confirm if that a mistake on my implementation. 

Thanks
